### PR TITLE
feat: add controller to as-controller-board-status

### PR DESCRIPTION
**Pull request by vli11**
closes: issue #40
Signed-off-by: Valina Li <valina.li@intel.com>

### CODE REVIEW

File: as-controller-board-status/routes/controller.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	return Controller{
+		lc:          lc,
+		service:     service,
+		boardStatus: boardStatus,
+	}
+}
+
+func (c *Controller) AddAllRoutes() error {
+	// Add the "status" REST API route
+	err := c.service.AddRoute("/status", c.GetStatus, "GET", "OPTIONS")
```
Comment (jim-wang-intel) [inline as-controller-board-status/routes/controller.go:33]: http.MethodGet, http.MethodOptions for constants

File: as-controller-board-status/main.go [hunk 2]
```diff
@@ -31,20 +32,41 @@ func main() {
...
+	if subscriptionClient == nil {
+		lc.Errorf("error notification service missing from client's configuration")
+		os.Exit(1)
 	}
 
+	config := new(functions.ControllerBoardStatusAppSettings)
 	// Retrieve & parse the required application settings into a proper
 	// configuration struct
-	err := utilities.MarshalSettings(appSettings, boardStatus.Configuration, false)
+	err := utilities.MarshalSettings(appSettings, config, false)
```
Comment (lenny-intel) [inline as-controller-board-status/main.go:44]: We can replace use of `utilities.MarshalSettings` with use of custom configuration. Please create an issue.

File: as-controller-board-status/main.go [hunk 3]
```diff
@@ -31,20 +32,41 @@ func main() {
...
 		os.Exit(1)
 	}
 
+	boardStatus := functions.CheckBoardStatus{
+		DoorClosed:         true, // Set default door state to closed
+		Configuration:      config,
+		SubscriptionClient: subscriptionClient,
+	}
+
+	// Subscribe to the EdgeX notification service
```
Comment (lenny-intel) [inline as-controller-board-status/main.go:56]: Remove command since this is obvious from the code.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: as-controller-board-status/functions/output.go [hunk 4]
```diff
@@ -262,16 +260,11 @@ func (boardStatus *CheckBoardStatus) processVendingDoorState(lc logger.LoggingCl
...
-func GetStatus(writer http.ResponseWriter, req *http.Request) {
-	utilities.ProcessCORS(writer, req, func(writer http.ResponseWriter, req *http.Request) {
-		controllerBoardStatusJSON, err := utilities.GetAsJSON(controllerBoardStatus)
-		if err != nil {
-			utilities.WriteStringHTTPResponse(writer, req, http.StatusInternalServerError, "Failed to serialize the controller board's current state.", true)
-			return
-		}
+func (boardStatus *CheckBoardStatus) GetControllerBoardStatus() ControllerBoardStatus {
+	return controllerBoardStatus
+}
```
Comment (jim-wang-intel) [inline as-controller-board-status/functions/output.go:265]: Is this necessary?  if controllerBoardStatus is a global variable, why could we directly access it?

File: as-controller-board-status/functions/output.go [hunk 5]
```diff
@@ -262,16 +260,11 @@ func (boardStatus *CheckBoardStatus) processVendingDoorState(lc logger.LoggingCl
...
 
-		utilities.WriteJSONHTTPResponse(writer, req, http.StatusOK, controllerBoardStatusJSON, false)
-	})
+// for testing purpose
+func (boardStatus *CheckBoardStatus) SetControllerBoardStatus(controllerBoardStatusInput ControllerBoardStatus) {
+	controllerBoardStatus = controllerBoardStatusInput
 }
```
Comment (jim-wang-intel) [inline as-controller-board-status/functions/output.go:270]: same comment above

**Review by jim-wang-intel — CHANGES_REQUESTED** [review]

File: as-controller-board-status/main.go (hunk 2 above)
Comment (vli11) [reply as-controller-board-status/main.go:44]: issue #44 created

File: as-controller-board-status/functions/output.go (hunk 4 above)
Comment (lenny-intel) [reply as-controller-board-status/functions/output.go:265]: Because it is global to this package, not the others...

Comment (vli11) [reply as-controller-board-status/functions/output.go:265]: controllerBoardStatus is not a global variable

File: as-controller-board-status/functions/output.go (hunk 5 above)
Comment (vli11) [reply as-controller-board-status/functions/output.go:270]: not a global variable

File: as-controller-board-status/functions/output.go (hunk 4 above)
Comment (lenny-intel) [reply as-controller-board-status/functions/output.go:265]: local globals like `controllerBoardStatus` are bad practice.
We should move `controllerBoardStatus ` into the `CheckBoardStatus` struct and not have it as package global.

Comment (jim-wang-intel) [reply as-controller-board-status/functions/output.go:265]: I would like to have discussion about this please

File: as-controller-board-status/functions/output.go (hunk 5 above)
Comment (jim-wang-intel) [reply as-controller-board-status/functions/output.go:270]: unit test is the same package as output.go, on the same package so it is global to unit tests.

**Review by jim-wang-intel — CHANGES_REQUESTED** [review]
please resolve my concerns thanks

File: as-controller-board-status/functions/common.go [hunk 6]
```diff
@@ -13,11 +13,11 @@ import (
...
 const (
 	// RESTPost is a const used for REST commands using the specified method.
-	RESTPost = "POST"
+	RESTPost = http.MethodPost
 	// RESTPut is a const used for REST commands using the specified method.
-	RESTPut = "PUT"
+	RESTPut = http.MethodPut
 	// RESTGet is a const used for REST commands using the specified method.
-	RESTGet = "GET"
+	RESTGet = http.MethodGet
```
Comment (lenny-intel) [inline as-controller-board-status/functions/common.go:20]: Just remove these constants and use the ones from `http.` in place of them

File: as-controller-board-status/functions/output.go (hunk 5 above)
Comment (lenny-intel) [reply as-controller-board-status/functions/output.go:270]: > unit test is the same package as output.go, on the same package so it is global to unit tests.

@jim-wang-intel, If it is moved into the struct then it resolves the "global" issue. :-)

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: as-controller-board-status/functions/output.go (hunk 5 above)
Comment (vli11) [reply as-controller-board-status/functions/output.go:270]: issue #46 created to resolve this "global" issue

Comment (jim-wang-intel) [reply as-controller-board-status/functions/output.go:270]: yes, agreed, just discussed with @vli11 and we agreed to create a separate PR to resolve this.

**Review by jim-wang-intel — APPROVED** [review]
all my issues addressed - thanks

File: as-controller-board-status/functions/output.go (hunk 4 above)
Comment (vli11) [reply as-controller-board-status/functions/output.go:266]: issue #46 created to track this change to be done

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**